tests/unit/defrag.cpp

```diff
@@ -0,0 +1,522 @@
+ *    State: data.0(sorted with view, 50 records), data.1(unsorted, 100 records)
+ * 5) Check that bases contains all 150 records
+ */
+BOOST_AUTO_TEST_CASE( test_first_base_sorted_second_base_unsorted ) {
+	const size_t RECORDS_IN_BLOB = 100;
+	const size_t TOTAL_RECORDS = 2 * RECORDS_IN_BLOB;
+	const size_t RECORDS_TO_REMOVE = RECORDS_IN_BLOB / 2;
+	std::mt19937 gen(42);
+	std::uniform_int_distribution<unsigned> dist;
+	auto gen_random_data = [&gen, &dist]() {
+		return generate_random_data(gen, dist);
+	};
+
+	eblob_wrapper wrapper;
+	auto config = wrapper.default_config;
+	config.blob_size = 10 * (1ULL << 30); // 10Gib
+	config.records_in_blob = RECORDS_IN_BLOB;
+	config.defrag_timeout = 0; // we don't want to autodefrag
+	config.defrag_time = 0;
+	config.defrag_splay = 0;
+
+	wrapper.start(&config);
+	BOOST_REQUIRE(wrapper.get() != nullptr);
+
+	std::vector<Item> shadow_elems;
+	int element_counter = 0;
+
+	// Insert items to form two full bases
+	for (size_t index = 0; index != TOTAL_RECORDS; ++index) {
+		shadow_elems.push_back(wrapper.generate_item(element_counter, gen_random_data()));
+		++element_counter;
+		BOOST_REQUIRE_EQUAL(wrapper.insert_item(shadow_elems.back()), 0);
+	}
+
+	BOOST_REQUIRE_EQUAL(datasort(wrapper, {0}), 0);
+
+	// Remove a half items from first base
+	for (size_t index = 0; index != RECORDS_TO_REMOVE; ++index) {
+		BOOST_REQUIRE_EQUAL(wrapper.remove_item(shadow_elems[index]), 0);
+	}
+
+	shadow_elems = std::vector<Item>(shadow_elems.begin() + RECORDS_TO_REMOVE, shadow_elems.end());
+	std::sort(shadow_elems.begin(), shadow_elems.begin() + (RECORDS_IN_BLOB - RECORDS_TO_REMOVE));
+
+	// TODO: нужно проверить, что view чанки добавились.
+	BOOST_REQUIRE_EQUAL(datasort(wrapper, {0}), 0);
+
+	struct eblob_index_block full_range; // можно не указывать
+	memset(full_range.start_key.id, 0, EBLOB_ID_SIZE);
+	memset(full_range.end_key.id, 0xff, EBLOB_ID_SIZE);
+
+	iterator_private priv(shadow_elems);
+	struct eblob_iterate_callbacks callbacks;
+	memset(&callbacks, 0, sizeof(callbacks));
+	callbacks.iterator = iterate_callback;
+	struct eblob_iterate_control eictl;
+	memset(&eictl, 0, sizeof(eictl));
+	eictl.b = wrapper.get();
+	eictl.log = config.log;
+	eictl.flags = EBLOB_ITERATE_FLAGS_ALL | EBLOB_ITERATE_FLAGS_READONLY;
+	eictl.iterator_cb = callbacks;
+	eictl.range = &full_range;
+	eictl.range_num = 1;
+	eictl.priv = &priv;
+
+	eblob_iterate(wrapper.get(), &eictl);
+	BOOST_REQUIRE_EQUAL(priv.number_checked, TOTAL_RECORDS - RECORDS_TO_REMOVE);
+}
+
+
+/**
+ * 1) Make two bases with 100 records each.
+ *    State: data.0(unsorted, 100 records), data.1(unsorted, 100 records)
+ * 2) Defrag first base
+ *    State: data.0(sorted, 100 records), data.1(unsorted, 100 records)
+ * 3) Remove a half from each base
+ *    State: data.0(sorted, 50 records), data.1(unsorted, 50 records)
+ * 4) Defrag two bases with view on first base
+ *    State: data.0(sorted, 100 records)
+ * 5) Check that result base contains 100 records
+ */
```

Comment:
All the test cases are:
* Setup: N bases (each base has some properties)
* Run the action that is tested (defrag)
* Check that all data is left at that: No loose of items, no leaks of removed items.

The first and the last step can be implemented as common code (parametrized a little).

As about the last step: you have opportunity to check not only count of items, but also their keys and their data. It must be done.

Reply:
> you have opportunity to check not only count of items, but also their keys and their data. It must be done.
currently it's done [here](https://github.com/interiorem/eblob/blob/eb10a9259c1b19a4bdd65c16f7f5008d1a1c6000/tests/unit/defrag.cpp#L78-L90)

Reply:
Really =) Ok.

So the issue left with grouping code.